File: profiling/runner_base.py
import argparse
import csv
import logging
import time

import utils.profile as memory_profiler
import utils.constants as constants

def main(language: str, experiment: str, loop: int) -> None:
    with open(f"{constants.LOG_PATH}/{language}_{experiment}_memory.csv", mode='w') as file:
        writer = csv.writer(file)
        writer.writerow(["base", "tmux"])
        
        for i in range(loop):
            logging.info(f"Loop: {i+1}/{loop}")
            memory_page = []
            
            # Base
            profiler = memory_profiler.MemoryProfiler(language, experiment, f"echo 'Hello'", f"echo 'Hello'")
            profiler.gracefuly_stop_fc_vm()
            profiler.start_new_vm()
            time.sleep(3)
            
            profiler.take_snapshot()
            profiler.load_snapshot()
            memory_page.append(profiler.get_stable_memory_page(120, 15))
            logging.info(f"Server memory page: {memory_page[0]}")
            
            profiler.gracefuly_stop_fc_vm()
            
            # Tmux
            profiler = memory_profiler.MemoryProfiler(language, experiment, f"echo 'Hello'", f"echo 'Hello'")
            profiler.gracefuly_stop_fc_vm()
            profiler.start_new_vm()
            
            profiler.run_server()
            profiler.take_snapshot()
            profiler.load_snapshot()
            memory_page.append(profiler.get_stable_memory_page(120, 15))
            logging.info(f"Server memory page: {memory_page[1]}")
            
            profiler.gracefuly_stop_fc_vm()
            
            writer.writerow(memory_page)


if __name__ == '__main__':
    logging.basicConfig(level=logging.INFO)
    parser = argparse.ArgumentParser(description='Profile the code')
    parser.add_argument('--language', '-l', type=str , help='Program language')
    parser.add_argument('--expr', '-e', type=str, help='Experiment name')
    parser.add_argument('--loop', '-n', type=int, help='Number of experiment loop')
    args = parser.parse_args()
    
    main(args.language, args.expr, args.loop)

File: go_profiling/runner_base.go
package main

import (
	"encoding/csv"
	"flag"
	"fmt"
	"os"
	"strconv"
	"time"

	log "github.com/sirupsen/logrus"

	"github.com/JooyoungPark73/fc_benchmark/go_profiling/utils"
)

// Main function for profiling
func RunBase() {
	// Command-line arguments parsing
	language := flag.String("language", "", "Program language")
	experiment := flag.String("expr", "", "Experiment name")
	loop := flag.Int("loop", 1, "Number of experiment loop")
	flag.Parse()

	// Open CSV file for writing
	filePath := fmt.Sprintf("%s/%s_%s_memory.csv", utils.Log_Path, *language, *experiment)
	file, err := os.Create(filePath)
	if err != nil {
		log.Fatalf("Failed to create CSV file: %v", err)
	}
	defer file.Close()

	writer := csv.NewWriter(file)
	defer writer.Flush()

	// Write CSV header
	writer.Write([]string{"base", "tmux"})

	// Loop through experiments
	for i := 0; i < *loop; i++ {
		log.Infof("Loop: %d/%d\n", i+1, *loop)
		memoryPage := []string{}

		// Base case
		profiler, _ := utils.NewMemoryProfiler(*language, *experiment, "echo 'Hello'", "echo 'Hello'")
		profiler.GracefullyStopFCVM()
		profiler.StartNewVM()
		time.Sleep(3 * time.Second)

		profiler.TakeSnapshot()
		profiler.LoadSnapshot()
		memoryPage = append(memoryPage, strconv.Itoa(profiler.GetStableMemoryPage(120, 15)))
		log.Infof("Server memory page: %s\n", memoryPage[0])

		profiler.GracefullyStopFCVM()

		// Tmux case
		profiler, _ = utils.NewMemoryProfiler(*language, *experiment, "echo 'Hello'", "echo 'Hello'")
		profiler.GracefullyStopFCVM()
		profiler.StartNewVM()

		profiler.RunServer()
		profiler.TakeSnapshot()
		profiler.LoadSnapshot()
		memoryPage = append(memoryPage, strconv.Itoa(profiler.GetStableMemoryPage(120, 15)))
		log.Infof("Server memory page: %s\n", memoryPage[1])

		profiler.GracefullyStopFCVM()

		// Write memory pages to CSV
		writer.Write(memoryPage)
	}
}
